### ISSUE TITLE
[BUG] Many CVES in kube-vip-iptables

### ISSUE THREAD

**Issue by cwayne18** [harvester/harvester#6904]
In RKE2, we bundle the kube-vip-iptables tarball as its needed by the harvester chart, but the version is quite a bit out of date (0.6.0 -> 0.8.4).

**Comment by bk201** [comment]
cc @starbops to check, is this the vip in guest clusters?

**Comment by starbops** [comment]
Yes, kube-vip as a dependency to harvester-cloud-provider in downstream clusters is still in 0.6.0. We need to bump it too.

**Comment by starbops** [comment]
## Test Plan

1. Prepare an external DHCP server with appropriate configuration (in order to serve the LB-type Service created in the downstream cluster later)
2. Prepare a Harvester cluster in v1.4.0
   1. Create an image, e.g., https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-amd64.img
   2. (Optional) Create a Cluster Network and its Network Config
   3. Create a Virtual Machine Network associated with the Cluster Network (if no new Cluster Network is created, select the `mgmt` one instead)
4. Import it to a Rancher Manager
5. From the Rancher dashboard, go into the imported Harvester's page and then create an IPPool for Cluster-type load balancers
6. Create a downstream RKE2 cluster with the image and network created above (made sure the Harvester cloud provider is selected)
7. After the downstream cluster is ready, download the kubeconfig file and use it for the following commands by `export KUBECONFIG=<kubeconfig>`.
8. Upgrade the harvester-cloud-provider chart with the following commands
   ```
   # For reviewers
   git clone https://github.com/starbops/harvester-charts.git
   cd charts/
   git checkout fix-6904
   cd charts/harvester-cloud-provider/
   helm upgrade --install harvester-cloud-provider ./ --reset-then-reuse-values -n kube-system

   # For QAs (after the PR is merged and the new version of harvester-cloud-provider is released)
   helm repo add harvester https://charts.harvesterhci.io
   helm repo update
   helm upgrade --install harvester-cloud-provider harvester/harvester-cloud-provider --version=<tbd-version> --reset-then-reuse-values -n kube-system
   ```
9. On the downstream cluster, create an `nginx` Deployment along with an LB-type Service with the `Pool` IPAM selected
10. Create another `nginx` Deployment along with a LB-type Service with the `DHCP` IPAM selected
11. Ensure the two LB-type Services get their IP address from the IPPool and DHCP server, respectively
12. Test the connection to the above two LB IPs works (you should be able to see the "Welcome to NGINX" page)
13. Delete the two LB-type Services
14. Test the connection to the two LB IPs. The connections should fail.

**Comment by m-ildefons** [comment]
@cwayne18 this issue seems to be fixed by virtue of the kube-vip dependency having been updated in the harvester cloud provider chart and the harvester cloud provider in turn having been updated in the rancher/charts repo.
However I'm having a hard time figuring out what versions of Rancher are now affected by this.
Is it correct that the fix is independent of the Harvester and Rancher versions and that it's automatically delivered to all freshly set up guest clusters (i.e. any test environment that I set up today should have it)? Or is this fix only delivered through selected Rancher versions/to selected downstream cluster versions?

Besides this, I wonder if there shouldn't be a more reliable and scalable mechanism to keep our dependencies up to date. By now there is an even newer version of kube-vip available.

**Comment by starbops** [comment]
Hi @m-ildefons 

IIRC, if a downstream K8s cluster is created with the `Harvester` cloud provider selected, the HCP version installed depends on which Kubernetes version is chosen and with which version the Rancher Manager is. For instance, if the downstream cluster is of version `v1.30.9+rke2r1` and the Rancher Manager is `v2.9.4`, the following command can help finding the bundled HCP version:

```
$ curl -sSfL https://releases.rancher.com/kontainer-driver-metadata/release-v2.9/data.json | jq '.rke2.releases[] | select(.version=="v1.30.9+rke2r1") | .charts."harvester-cloud-provider".version'
"0.2.900"
```

However, for the `v1.30.8+rke2r1` downstream cluster provisioned with the same Rancher Manager, the HCP version will be:

```
$ curl -sSfL https://releases.rancher.com/kontainer-driver-metadata/release-v2.9/data.json | jq '.rke2.releases[] | select(.version=="v1.30.8+rke2r1") | .charts."harvester-cloud-provider".version'
"0.2.600"
```

Later, users can choose to upgrade HCP from the Rancher marketplace (if there's one available). The HCP version will look like `10x.0.y+up0.2.9`. This is essential HCP 0.2.9.

Regarding your question, I would say the fix is independent of the Harvester version but somewhat dependent on Rancher version and the downstream cluster's Kubernetes version.

> Besides this, I wonder if there shouldn't be a more reliable and scalable mechanism to keep our dependencies up to date. By now there is an even newer version of kube-vip available.

I have some thoughts about this. The development work around kube-vip is still in a fast pavement, and sometimes breaking changes or bugs are introduced over time. I'm afraid that it's inevitable that we need to trial all the candidate versions and hand-pick one to bump to. Ultimately, it's our responsibility to ensure the dependency bump does not break things. Before we have a well-written integration/e2e testing that exists for the Harvester cloud provider, we need to rely on human effort. Just my 2c.

**Comment by khushboo-rancher** [comment]
It does depend on rke2 versions and chart wise, the change will be available on all the active Rancher versions. More details are below:

**Chart updates in App:**
1. Code is merged in dev branches of active Rancher versions. e.g. dev-v2.10, dev-v2.9 etc
2. On releasing them, they are available on `release-v2.9`, `release-v2.10` etc branches.
3. On a released Rancher version, app is pointing to `release-v2.x` branch. So, user will get the changes as soon as the charts are released.
4. In this case, all the Rancher version 2.9.0, 2.9.1, 2.9.2, so on, 2.10.0+ will get the update.

**Chart updates in RKE2**
1. Code is merged in rancher/rke2-charts and rancher/rke2 repo.
2. Every month, RKE2 releases patch versions of 4 maintained rke2 versions. e.g. v1.32.2+rke2r1, v1.31.5+rke2r1, v1.30.9+rke2r1, v1.29.8+rke2r1
3. Once above versions are out, the updated charts become available on these versions and above.

-----

This issue is ok to close as we have already validated the cloud version bump to v0.2.9 with dependency kube-vip 0.6.4
The chart is available on Rancher 2.8, 2.9, 2.10 app and RKE2 guest cluster versions v1.31.5+rke2r1, v1.30.9+rke2r1, v1.29.13+rke2r1 and above.

Testing details - https://github.com/harvester/harvester/issues/7333#issuecomment-2705030411

cc: @harvester/qa
